File: go/internal/tools/linode_kernels.go
package tools

import (
	"context"
	"fmt"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/chadit/LinodeMCP/go/internal/config"
	linodev1 "github.com/chadit/LinodeMCP/go/internal/genpb/linode/mcp/v1"
	"github.com/chadit/LinodeMCP/go/internal/linode"
	"github.com/chadit/LinodeMCP/go/internal/profiles"
	"github.com/chadit/LinodeMCP/go/internal/toolschemas"
)

const (
	kernelsPageSizeMin    = 25
	kernelsPageSizeMax    = 500
	kernelIDPrefixLinode  = "linode"
	errKernelIDIdentifier = "kernel_id must be a kernel identifier like linode/latest-64bit"
)

// NewLinodeKernelListTool creates a tool for listing Linode kernels.
func NewLinodeKernelListTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newProtoListToolPaginated(
		cfg,
		"linode_kernel_list",
		"Lists available Linode kernels with optional pagination.",
		"Page of results to return (optional, minimum 1).",
		"Number of results per page (optional, 25-500).",
		func(ctx context.Context, client *linode.Client, page, pageSize int) ([]*linodev1.Kernel, error) {
			return client.ListKernelsProto(ctx, page, pageSize)
		},
		kernelsPaginationFromTool,
		nil,
		kernelListResponse,
	)

	return tool, profiles.CapRead, handler
}

func kernelListResponse(items []*linodev1.Kernel, count int32, filter *string) *linodev1.KernelListResponse {
	return &linodev1.KernelListResponse{Count: count, Filter: filter, Kernels: items}
}

// NewLinodeKernelGetTool creates a tool for retrieving one Linode kernel.
func NewLinodeKernelGetTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool := mcp.NewToolWithRawSchema(
		"linode_kernel_get",
		"Gets one Linode kernel by ID.",
		toolschemas.Schema("linode.mcp.v1.KernelGetInput"),
	)

	handler := func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		return handleKernelGetRequest(ctx, &request, cfg)
	}

	return tool, profiles.CapRead, handler
}

func handleKernelGetRequest(ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config) (*mcp.CallToolResult, error) {
	kernelID, ok := getKernelIDArg(request)
	if !ok {
		return mcp.NewToolResultError(errKernelIDIdentifier), nil
	}

	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	kernel, err := client.GetKernelProto(ctx, kernelID)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to retrieve kernel: %v", err)), nil
	}

	return MarshalProtoToolResponse(kernel)
}

func kernelsPaginationFromTool(request *mcp.CallToolRequest) (int, int, string) {
	args := request.GetArguments()

	page, validationMessage := optionalPaginationInt(args, "page", 1, 0)
	if validationMessage != "" {
		return 0, 0, validationMessage
	}

	pageSize, validationMessage := optionalPaginationInt(args, "page_size", kernelsPageSizeMin, kernelsPageSizeMax)
	if validationMessage != "" {
		return 0, 0, validationMessage
	}

	return page, pageSize, ""
}

func getKernelIDArg(request *mcp.CallToolRequest) (string, bool) {
	raw, ok := request.GetArguments()["kernel_id"].(string)
	if !ok {
		return "", false
	}

	kernelID := strings.TrimSpace(raw)
	if kernelID == "" || strings.ContainsAny(kernelID, "?#") || strings.Contains(kernelID, "..") {
		return "", false
	}

	prefix, name, found := strings.Cut(kernelID, "/")
	if !found || prefix != kernelIDPrefixLinode || name == "" || strings.Contains(name, "/") {
		return "", false
	}

	return kernelID, true
}

File: python/src/linodemcp/tools/linode_kernels.py
"""Linode kernel tools."""

from __future__ import annotations

import re
from typing import TYPE_CHECKING, Any

from mcp.types import TextContent, Tool

from linodemcp.genpb.linode.mcp.v1 import kernel_pb2
from linodemcp.profiles import Capability
from linodemcp.tools.helpers import (
    error_response,
    execute_tool,
    pagination_int_argument,
)
from linodemcp.tools.proto_response import (
    serialize_api_response,
    serialize_list_response,
)
from linodemcp.tools.toolschemas import schema

if TYPE_CHECKING:
    from linodemcp.linode import RetryableClient


def create_linode_kernel_list_tool() -> tuple[Tool, Capability]:
    """Create the linode_kernel_list tool."""
    return Tool(
        name="linode_kernel_list",
        description="Lists available Linode kernels.",
        inputSchema=schema("linode.mcp.v1.KernelListInput"),
    ), Capability.Read


async def handle_linode_kernel_list(
    arguments: dict[str, Any], cfg: Any
) -> list[TextContent]:
    """Handle linode_kernel_list tool request."""
    try:
        page = pagination_int_argument(arguments, "page", 1)
        page_size = pagination_int_argument(arguments, "page_size", 25, 500)
    except (TypeError, ValueError) as exc:
        return error_response(str(exc))

    async def _call(client: RetryableClient) -> dict[str, Any]:
        raw = await client.list_kernels(page=page, page_size=page_size)
        return serialize_list_response(
            raw,
            "kernels",
            kernel_pb2.KernelListResponse(),
        )

    return await execute_tool(cfg, arguments, "list Linode kernels", _call)


_KERNEL_ID_PATTERN = re.compile(r"^(?!.*\.\.)linode/[A-Za-z0-9._-]+$")


def _validated_kernel_id(value: object) -> tuple[str | None, str | None]:
    """Validate and normalize a Linode kernel ID path parameter."""
    if not isinstance(value, str) or not value.strip():
        return None, "kernel_id is required"
    kernel_id = value.strip()
    if not _KERNEL_ID_PATTERN.fullmatch(kernel_id):
        return None, "kernel_id must look like linode/latest-64bit"
    return kernel_id, None


def create_linode_kernel_get_tool() -> tuple[Tool, Capability]:
    """Create the linode_kernel_get tool."""
    return Tool(
        name="linode_kernel_get",
        description="Gets a single Linode kernel by ID.",
        inputSchema=schema("linode.mcp.v1.KernelGetInput"),
    ), Capability.Read


async def handle_linode_kernel_get(
    arguments: dict[str, Any], cfg: Any
) -> list[TextContent]:
    """Handle linode_kernel_get tool request."""
    kernel_id, kernel_id_err = _validated_kernel_id(arguments.get("kernel_id"))
    if kernel_id_err is not None or kernel_id is None:
        return error_response(kernel_id_err or "kernel_id is required")

    async def _call(client: RetryableClient) -> dict[str, Any]:
        kernel = await client.get_kernel(kernel_id)
        return serialize_api_response(kernel, kernel_pb2.Kernel())

    return await execute_tool(cfg, arguments, "retrieve Linode kernel", _call)
